File: 정은기/최단 경로(Dijkstra)/BOJ_11265_끝나지_않는_파티.java
package aim_to_platinum.week06_shortest_path.b11265;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

/*
1. 문제 요약
- 모든 파티장끼리의 최단 경로를 구해야 하므로 플로이드 알고리즘 사용
    - 파티장(이하 노드)간 인접행렬 입력받아 중간 노드 거치게 만들어 최단거리 업데이트
- 이후 서비스 요청 수만큼 출발->도착 최단경로 값과 시간 비교
- 노드 수를 N, 서비스 요청 손님 수를 M 이라고 할 때,
    N(5 ≤ N ≤ 500), M(1 ≤ M ≤ 10,000) 이고
    인접행렬을 board 라고 하면 board[i][j] 안에 들어갈 수 있는 수 T의 범위는
    T(1 ≤ T ≤ 1,000,000)

2. 아이디어 (문제 접근법)
[아이디어-1]
- 플로이드 알고리즘 사용하여 모든 정점간 최단 경로 확인
- 이후 board[from][to] <= weight 일 경우 "Enjoy other party",
    아닐경우 "Stay here" 출력

3. 어려움 및 해결
-
 */
public class BOJ_11265_끝나지_않는_파티 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer stk;

        stk = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(stk.nextToken());
        int M = Integer.parseInt(stk.nextToken());
        long[][] board = new long[N][N];

        for(int i=0; i<N; i++){
            stk = new StringTokenizer(br.readLine());
            for(int j=0; j<N; j++){
                board[i][j] = Long.parseLong(stk.nextToken());
            }
        }

        for(int k=0; k<N; k++){
            for(int i=0; i<N; i++){
                for(int j=0; j<N; j++){
                    if(board[i][j] > board[i][k] + board[k][j]){
                        board[i][j] = board[i][k] + board[k][j];
                    }
                }
            }
        }

        for(int i=0; i<M; i++){
            stk = new StringTokenizer(br.readLine());
            int from = Integer.parseInt(stk.nextToken()) - 1;
            int to = Integer.parseInt(stk.nextToken()) - 1;
            long weight = Long.parseLong(stk.nextToken());

            if(board[from][to] <= weight){
                System.out.println("Enjoy other party");
            }else{
                System.out.println("Stay here");
            }
        }
    }

}

File: 이다인/백준/최단 경로(Dijkstra)/BOJ_11265_끝나지_않는_파티.py
'''
1. 문제요약
- A 파티장에서 B 파티장으로 직접 연결된 일방통행 도로를 만들었지만, 다른 파티장을 경유해서 더 빨리 갈 수 있는 경우가 있을 수 있음
- B 파티장에서 C 시간 후에 다음 파티가 열릴 때, A 파티장에서 시간내에 도착할 수 있는지 없는지 알아봐주는 서비스

2. 아이디어(문제접근법)
- 서비스를 요청한 손님이 최대 10,000명이므로 모든 파티장에서 다른 모든 파티장으로의 이동 시간을 먼저 구함
- 플로이드 워셜 알고리즘을 활용해서 계산한 후 graph 배열에 저장
- graph[a][b]가 c보다 작거나 같으면 시간내에 도착할 수 있으므로 "Enjoy other party" 출력, 아니면 "Stay here" 출력
'''

import sys

input = sys.stdin.readline
n, m = map(int, input().split())
graph = [[0] * (n + 1)]

for _ in range(n):
    graph.append([0] + list(map(int, input().split())))

for k in range(1, n + 1):
    for a in range(1, n + 1):
        for b in range(1, n + 1):
            graph[a][b] = min(graph[a][k] + graph[k][b], graph[a][b])

for _ in range(m):
    a, b, c = map(int, input().split())
    if graph[a][b] <= c:
        print("Enjoy other party")
    else:
        print("Stay here")